keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,366 @@
+As part of the investigation, it may be necessary to scrutinize the relevant logs of each component in order to figure out the series of cluster processing.
+It takes long time to scrutinize the relevant logs without this log tracking feature, because component logs are independent of each other, and it is difficult to find related logs and link them.
+
+This is similar to the [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/), except for the following points.
+
+ - Audit only collects information about http request sending and receiving in kube-apiserver, so it can't track internal work of each component.
+ - Audit logs can't be associated to logs related to user operation (kubectl operation), because auditID is different for each http request.
+
+#### Story 2
+
+Failed to attach PV to pod
+Prerequisite: It has been confirmed that the PV has been created successfully.
+In this case, the volume generation on the storage side is OK, and there is a possibility that the mount process to the container in the pod is NG.
+In order to identify the cause, it is necessary to look for the problem area while checking the component (kubelet) log as well as the system side syslog and mount related settings.
+
+This log tracking feature is useful to identify the logs related to specific user operation  and cluster processing, and can reduce investigation cost in such cases.
+### Summary of Cases
+
+ - Given a component log(such as error log), find the API request that caused this (error) log.
+ - Given an API Request(such as suspicious API request), find the resulting component logs.
+
+### Logging metadata
+
+We use three logging meta-data, and propagate them each K8s component by using OpenTelemetry.
+OpenTelemetry has SpanContext which is used for propagation of K8s component.
+
+| meta-data name | feature |
+| ------ | ------ |
+| trace-id | We use SpanContext.TraceID as trace-id<br>trace-id spans an user request.<br>trace-id is unique for user's request |
+| span-id | We use SpanContext.SpanID as span-id<br>span-id spans a controller action.<br>span-id is unique for controller action |
+| initial-trace-id | We implement new id(InitialTraceID) to SpanContext<br>We use SpanContext.InitialTraceID as initial-trace-id<br>initial-trace-id spans the entire object lifecycle. <br> initial-trace-id is unique for related objects |
+
+All of three id's inception is from object creation and it dies with object deletion
+
+## Design Details
+
+### Prerequisite
+
+We need to consider three cases:
+- Case1: Requests from kubectl that creating an object
+- Case2: Requests from kubectl other than creating (e.g. updating, deleting) an object
+- Case3: Requests from controllers
+
+The design below is based on the above three cases
+
+The following picture show our design![design](./overview.png)
+
+
+
+we don't have any modifications in kubectl in this design.
+
+### Design of ID propagation in apiserver
```

Comment:
I thought we weren't going to do any changes to the API Server as part of this KEP?

Edit: I see... I think we can avoid these changes by using baggage.

Reply:
This still says we are adding propagation in the API server.  I don't think we want to do that.

Reply:
@dashpole Here is a [k8s demo](https://github.com/Hellcatlk/kubernetes/pull/2) and [webhook demo](https://github.com/Hellcatlk/mutating-trace-admission-controller/blob/trace-noapiserver/pkg/util/trace/trace.go).

In the demo, we avoid modifying apiserver, by the following points:

1. Do `StarSpan` in webhook.
2. When RS creates Pod, we copy InitialTraceID and SpanContext from RS's annotations to Pod's annotations.

Will the above two points be allowed?

Reply:
If we don't do changes in apiserver, the TraceContext and Baggage in http header cannot propagate to webhook, right ?
since the incomming http header will be consumed by the origin handlers chain(no handlers help put them to the r.ctx which will pass to next phase),  then TraceContext and Baggage lost.

Reply:
This is already being done as part of https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing#tracing-api-requests.  Feel free to reference that, but we don't need to propose it here.

Reply:
thanks a lot, it's very helpful. i missed this KEP before.

Reply:
Hey @dashpole , May i know if your (KEP)[https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing#tracing-api-requests] will change the  [client-go](https://github.com/kubernetes/client-go) to inject the Spancontext and Baggage into http request header(something like [request](#design-of-id-progagation-in-client-go) in this KEP) ?

Reply:
Yes, it will.  I'm working on the implementation right now, and I hope to have it upstream in the next week or so.

Reply:
it helps a lot,  thank you very much.

Reply:
Hey @dashpole, i noticed that you have submit a [PR](https://github.com/kubernetes/kubernetes/pull/94942) to kubernetes. I'm going to adapt/rebase our demo basing on you PR if you don't mind.

Reply:
You are welcome to, although it may be less work to wait until it is merged :).